library/src/main/scala/scala/scalajs/macroImpls/UseAsMacros.scala

```diff
@@ -0,0 +1,304 @@
+     *  [[srcTpe]] is only used for error messages.
+     */
+    private def checkMembers(srcTpe: Type, exports: Map[String, List[JSMember]],
+        rawMembers: Iterable[(String, JSMember)]): Unit = {
+
+      for ((name, jsMem) <- rawMembers) {
+        exports.get(name).fold {
+          val msg = s"$srcTpe does not export a member named $name"
+          c.error(c.enclosingPosition, msg)
+        } { candidates =>
+          if (!hasConformingMembers(jsMem, candidates)) {
+            val msg = s"$srcTpe does not export " + jsMem.displayStr(name)
+            c.error(c.enclosingPosition, msg)
+          }
+        }
+      }
+    }
+
+    /** Whether the [[candidates]] together conform to [[jsMem]] */
+    private def hasConformingMembers(jsMem: JSMember,
+        candidates: List[JSMember]): Boolean = {
+
+      def find(mem: JSMember) = candidates.exists(_.conformsTo(mem))
+
+      def hasConformingOverloads = jsMem match {
+        case jsMem @ JSMethod(params, _) if params.last.isDefault =>
+          /* If the last parameter is a default parameter, the method can also
+           * be implemented by two methods, one with and one without the
+           * parameter. Note that this does not generalize to more parameters
+           * due to named parameters.
+           */
+          val pinit = params.init
+          val noLastDefault = pinit :+ params.last.copy(isDefault = false)
+
+          find(jsMem.copy(params = pinit)) &&
+          find(jsMem.copy(params = noLastDefault))
+
+        case _ => false
+      }
+
+      find(jsMem) || hasConformingOverloads
+    }
+
+    /** Members that a facade type defines */
+    private def rawJSMembers(tpe: Type): Iterable[(String, JSMember)] = {
+
+      def isValidMember(mem: Symbol) = (
+          !JSObjectAncestors(mem.owner) &&
+          !mem.isConstructor &&
+          mem.isMethod &&
+          !mem.asTerm.isParamWithDefault
+      )
+
+      for (mem <- tpe.members if isValidMember(mem)) yield {
+        val (name, explicit) = jsName(mem.asMethod)
+
+        // apply is a special case, since when it is not explicitly specified,
+        // it will result in an actual function application in JavaScript
+        // Scala defined types cannot support this --> fail
+        if (name == "apply" && !explicit) {
+          c.abort(c.enclosingPosition, s"$tpe defines an apply method. This " +
+            "cannot be implemented by any Scala exported type, since it " +
+            "would need to chain Function's prototype.")
+        }
+
+        (name, jsMemberFor(tpe, mem.asMethod))
+      }
+    }
+
+    /** Returns the JavaScript name of a method and whether it is specified
+     *  explicitly or not.
+     */
+    private def jsName(sym: MethodSymbol): (String, Boolean) = {
+      val annot = memberAnnotations(sym).find(annotIs(_, JSNameAnnotation))
+      val name = annot.flatMap(annotStringArg).getOrElse(defaultName(sym))
+      (name, annot.isDefined)
```

Comment:
Hmmm... I might actually go with the ADT, otherwise the error messages will be very bad (and in the future we'll need to not always fail, since two `js.Any`s can actually conform in their apply method).